group1/Monika_Lambeva_82017/BinTree/BinTree.h

```diff
@@ -0,0 +1,236 @@
+#ifndef BINTREE_H
+#define BINTREE_H
+#include<cassert>
+#include<fstream>
+
+template<class T>
+class BinTree
+{
+private:
+    struct BinTreeNode
+    {
+        T data;
+        BinTreeNode *left, *right;
+    };    
+    BinTreeNode *root;
+
+public:
+    BinTree();
+
+    void add(const T& x, const char* trace);
+    BinTreeNode* locate(const char* trace);
+    T get(const char* trace);
+
+    //ex.19.1
+    int count();
+    int countHelp(BinTreeNode *curr);
+
+    //ex.19.2
+    int countEvens();
+    //int countEvensHelp(BinTreeNode *curr);
+
+    //ex.19.3
+    int searchCount(bool(*pred)(const T&));
+    int searchCountHelp(BinTreeNode *curr, bool(*pred)(const T&));
+
+    //ex.19.4
+    int height();
+    int heightHelp(BinTreeNode *curr);
+
+    void toGv(std::ostream& out);
+    void toGvHelp (std::ostream& out, BinTreeNode *curr);
+};
+
+template<class T>
+BinTree<T>::BinTree():root(nullptr){}
+
+template<class T>
+void BinTree<T>::add(const T& x, const char* trace)
+{
+    if(!root)
+    {
+        assert (!trace[0]);
+        root = new BinTree<T>::BinTreeNode{x, nullptr, nullptr};
+        return;
+    }
+    BinTreeNode *curr = root;
+    while(trace[1])
+    {
+        assert(trace[0] == 'L' || trace[0] == 'R');
+        if(trace[0] == 'L')
+        {
+            curr = curr->left;
+        }
+        else
+        {
+            curr = curr->right;
+        }
+        assert(curr);
+        trace++;
+    }
+    if(trace[0] == 'L')
+    {
+        assert(!curr->left);
+        curr->left = new BinTree<T>::BinTreeNode{x,nullptr,nullptr};
+    }
+    else if(trace[0] == 'R')
+    {
+        assert(!curr->right);
+        curr->right = new BinTree<T>::BinTreeNode{x,nullptr,nullptr};
+    }
```

Comment:
Can we change the trace in order to skip this part and end at the right spot after the loop?